### PR TITLE
Add support for batch key rotation

### PR DESCRIPTION
**Pull request by poornas**
of objects encrypted previously with sse-s3 or sse-kms

## Motivation and Context
Allow key rotation of previously sse-s3/sse-kms encrypted objects via batch jobs

## How to test this PR?
with `mc batch start alias ./rotate.yaml`

cat ./rotate.yaml
```
keyrotate:
  apiVersion: v1
  bucket: bucket
  prefix: 
  encryption:
    type: sse-kms # valid values are sse-s3 and sse-kms
    key: my-new-keys2 # valid only for sse-kms
    # context: <new-kms-key-context> # valid only for sse-kms

  # optional flags based filtering criteria
  # for all objects
  # flags:
  #   filter:
  #     newerThan: "84h" # match objects newer than this value (e.g. 7d10h31s)
  #     olderThan: "80h" # match objects older than this value (e.g. 7d10h31s)
  #     # createdAfter: "2023-03-02T15:04:05Z07:00" # match objects created after "date"
  #     # createdBefore: "2023-03-02T15:04:05Z07:00" # match objects created before "date"
  #     tags:
  #       - key: "name"
  #         value: "pick*" # match objects with tag 'name', with all values starting with 'pick'
  #     metadata:
  #       - key: "content-type"
  #         value: "image/*" # match objects with 'content-type', with all values starting with 'image/'
  #     kmskey: "key-id" # match objects with KMS key-id (applicable only for sse-kms)
  #   notify:
  #     endpoint: "https://notify.endpoint" # notification endpoint to receive job status events
  #     token: "Bearer xxxxx" # optional authentication token for the notification endpoint
  #   retry:
  #     attempts: 10 # number of retries for the job before giving up
  #     delay: "500ms" # least amount of delay between each retry
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#3ed3927c-5b7b-4562-ae54-5cf1123fa1dd) |
<a name="3ed3927c-5b7b-4562-ae54-5cf1123fa1dd"></a>
### 16844-0e4bb4d/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.7:31602
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp 591be4c897f6:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 8 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 9 seconds
(5/14) Running awscli tests ... done in 1 minutes and 16 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 17 seconds
(8/14) Running minio-go tests ... done in 1 minutes and 3 seconds
(9/14) Running minio-java tests ... done in 32 seconds
(10/14) Running minio-js tests ... FAILED in 57 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 9,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:71:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:166:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:90:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:670:10) at endReadableNT (internal/streams/readable.js:1333:12) at processTicksAndRejections (internal/process/task_queues.js:82:21)"
}
(10/14) Running minio-py tests ... done in 2 minutes and 4 seconds
(11/14) Running s3cmd tests ... done in 16 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 31 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/batch-handlers.go [hunk 1]
```diff
@@ -399,14 +400,36 @@ const (
...
-			} else {
-				ri.RetryAttempts = batchReplJobDefaultRetries
+			ri.Version = int(version)
+			switch {
+			case job.Replicate != nil:
+				if job.Replicate.Flags.Retry.Attempts > 0 {
+					ri.RetryAttempts = job.Replicate.Flags.Retry.Attempts
+				} else {
+					ri.RetryAttempts = batchReplJobDefaultRetries
+				}
```
Comment (harshavardhana) [inline cmd/batch-handlers.go]:
```suggestion
                                ri.RetryAttempts = batchReplJobDefaultRetries
				if job.Replicate.Flags.Retry.Attempts > 0 {
					ri.RetryAttempts = job.Replicate.Flags.Retry.Attempts
				}
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :x: [more...](#80916873-fda1-4157-9fb6-0bf9bf3dd19e) |
<a name="80916873-fda1-4157-9fb6-0bf9bf3dd19e"></a>
### 16844-afe351e/mint-pools.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.5:31166
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp b7f60412f404:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 8 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 10 seconds
(5/14) Running awscli tests ... done in 2 minutes and 30 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 19 seconds
(8/14) Running minio-go tests ... done in 57 seconds
(9/14) Running minio-java tests ... done in 36 seconds
(10/14) Running minio-js tests ... FAILED in 2 minutes and 0 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"Versioning tests on a buckets: getObject, fGetObject, getPartialObject, putObject, removeObject with versionId support\"",
  "duration": 64683,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty. You must delete all versions in the bucket. at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:71:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:166:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:90:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:670:10) at endReadableNT (internal/streams/readable.js:1333:12) at processTicksAndRejections (internal/process/task_queues.js:82:21)"
}
(10/14) Running minio-py tests ... done in 2 minutes and 19 seconds
(11/14) Running s3cmd tests ... done in 17 seconds
(12/14) Running s3select tests ... done in 5 seconds
(13/14) Running versioning tests ... done in 3 minutes and 20 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
PTAL at the conflicts @poornas.

@aead need your help in reviewing this PR for any mistakes. Please take a look.

File: cmd/batch-handlers.go [hunk 2]
```diff
@@ -570,7 +571,7 @@ func (r *BatchJobReplicateV1) StartFromSource(ctx context.Context, api ObjectLay
 				ri.trackCurrentBucketObject(r.Target.Bucket, oi, success)
 				globalBatchJobsMetrics.save(job.ID, ri)
 				// persist in-memory state to disk after every 10secs.
-				logger.LogIf(ctx, ri.updateAfter(ctx, api, 10*time.Second, job.Location))
+				logger.LogIf(ctx, ri.updateAfter(ctx, api, 10*time.Second, job.Location, job.Type()))
```
Comment (harshavardhana) [inline cmd/batch-handlers.go]:
```suggestion
				logger.LogIf(ctx, ri.updateAfter(ctx, api, 10*time.Second, job))
```

We could pass the `job` directly?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/batch-rotate.go [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+		if strings.HasPrefix(strings.ToLower(k), ReservedMetadataPrefixLower) {
+			encMetadata[k] = v
+		}
+	}
+
+	if (sseKMS || sseS3) && r.Encryption.Type == ssekms {
+		if err = r.Encryption.Validate(); err != nil {
+			return err
+		}
+		newKeyID = r.Encryption.Key
```
Comment (aead) [inline cmd/batch-rotate.go]:
```
newKeyID = strings.TrimPrefix(r.Encryption.Key, crypto.ARNPrefix)
```

**Review by aead — APPROVED** [review]
LGTM

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**